torch/fx/experimental/fuser.py

```diff
@@ -0,0 +1,49 @@
+import torch.fx as fx
+from torch.nn.utils.fusion import fuse_conv_bn_eval
+from typing import Type, Dict, Any, Tuple, Iterable
+import torch
+import copy
+
+def _parent_name(target : str) -> Tuple[str, str]:
+    """
+    Splits a qualname into parent path and last atom.
+    For example, `foo.bar.baz` -> (`foo.bar`, `baz`)
+    """
+    *parent, name = target.rsplit('.', 1)
+    return parent[0] if parent else '', name
+
+# Works for length 2 patterns with 2 modules
+def matches_module_pattern(pattern: Iterable[Type], node: fx.Node, modules: Dict[str, Any]):
+    if len(node.args) == 0:
+        return False
+    nodes: Tuple[Any, fx.Node] = (node.args[0], node)
+    for expected_type, current_node in zip(pattern, nodes):
+        if not isinstance(current_node, fx.Node):
+            return False
+        if current_node.op != 'call_module':
+            return False
+        if not isinstance(current_node.target, str):
+            return False
+        if current_node.target not in modules:
+            return False
+        if type(modules[current_node.target]) is not expected_type:
+            return False
+    return True
+
+
+def fuse(model: torch.nn.Module) -> torch.nn.Module:
+    patterns = [(torch.nn.Conv1d, torch.nn.BatchNorm1d), (torch.nn.Conv2d, torch.nn.BatchNorm2d), (torch.nn.Conv3d, torch.nn.BatchNorm3d)]
+    model = fx.symbolic_trace(model)
+    modules = dict(model.named_modules())
+    new_graph = copy.deepcopy(model.graph)
+    for pattern in patterns:
+        for node in new_graph.nodes:
+            if matches_module_pattern(pattern, node, modules):
+                if len(node.args[0].users) > 1:  # Output of conv is used by other nodes
+                    continue
+                fused_conv = fuse_conv_bn_eval(modules[node.args[0].target], modules[node.target])
+                parent_name, name = _parent_name(node.args[0].target)
+                setattr(modules[parent_name], name, fused_conv)
+                node.replace_all_uses_with(node.args[0])
+                new_graph.erase_node(node)
+    return fx.GraphModule(dict(model.named_modules()), new_graph)
```

Comment:
Is this right? above we're mutating the `modules` object, but then passing in a new dict created from the original symbolically traced Module to the constructor. Does this work?

Reply:
So the way this works is that the modules object is just a pointer to the underlying module of the model - we don't actually care what's in the modules object itself. I was a bit confused by this (and thought that perhaps both this code and the quantization fuser were doing something wrong), but I think it's correct now.

So imagine you had a module with just a `conv1` in it, then `named_modules` would return something like
```
{
   '': Module(conv1: Conv2d(...))
   "conv1": Conv2d(...)
}
```

Basically, the reason the code looks for the parent and then indexes with the qualified name is that this provides a pointer to the actual module in the model, and not that we're modifying the module class itself.